Prefix federation ID to e-cash notes when spending/reissuing/validating

shaurya947 opened the pull request:
We create a new struct called `OOBNotes` that encapsulates the federation ID + notes. Modules aren't aware of the federation ID so we also have to inject that.

Should close #2783

elsirion commented on modules/fedimint-mint-client/src/lib.rs:
```diff
@@ -74,14 +74,27 @@ const MINT_BACKUP_RESTORE_OPERATION_ID: OperationId = OperationId([0x01; 32]);
 pub const LOG_TARGET: &str = "client::module::mint";
 
+#[derive(Clone, Debug, Decodable, Encodable, Deserialize, Serialize)]
+pub struct OOBNotes {
+    federation_id: FederationId,
+    pub notes: TieredMulti<SpendableNote>,
+}
```
Why is one `pub` but not the other?

shaurya947 replied:
I initially exposed `notes` for `totalAmount()` but then decided to make a pass-through function on `OOBNotes` itself. I suppose they could both be non-`pub` and there could be getters for both fields? Or they could both be made `pub`? Probably the latter is good enough?

dpc replied:
I'd make both `pub`. This struct is just a no-logic aggregator of two datums.

elsirion commented on modules/fedimint-mint-client/src/lib.rs:
```diff
@@ -74,14 +74,27 @@ const MINT_BACKUP_RESTORE_OPERATION_ID: OperationId = OperationId([0x01; 32]);
 
 pub const LOG_TARGET: &str = "client::module::mint";
 
+#[derive(Clone, Debug, Decodable, Encodable, Deserialize, Serialize)]
```
The `Serialize`/`Deserialize` impl does not what we want here, the struct should be encoded as one string if the encoding format is human readable. If you aren't familiar with custom implementations of `Serialize`/`Deserialize` you could just keep the `serde_ecash` module around for now.

shaurya947 replied:
Ah I think I follow. We'd still like for the serialization of the notes (`OOBNotes`) to be one string containing a base-64 byte encoding of the struct. I will write a custom serializer.

dpc replied:
There's a `serde_as_encodable_hex` for exactly this already.

elsirion commented on modules/fedimint-mint-client/src/lib.rs:
```diff
@@ -1274,49 +1307,22 @@ pub struct SpendableNote {
 /// Base64 encode a set of e-cash notes. See also [`parse_ecash`].
-pub fn serialize_ecash(ecash: &TieredMulti<SpendableNote>) -> String {
+pub fn serialize_ecash(ecash: &OOBNotes) -> String {
     let mut bytes = Vec::new();
     Encodable::consensus_encode(ecash, &mut bytes).expect("encodes correctly");
     base64::encode(&bytes)
 }
 
 /// Decode a set of e-cash notes from a base64 string. See also
 /// [`serialize_ecash`].
-pub fn parse_ecash(s: &str) -> anyhow::Result<TieredMulti<SpendableNote>> {
+pub fn parse_ecash(s: &str) -> anyhow::Result<OOBNotes> {
```
These functions should be transformed into `FromStr` and `Display` implementations now that there is a local type for serializable e-cash.

elsirion commented on fedimint-cli/src/client.rs:
```diff
@@ -57,7 +57,7 @@ pub enum ClientCmd {
     /// Reissue notes received from a third party to avoid double spends
     Reissue {
         #[clap(value_parser = parse_ecash)]
```
Once `OOBNotes` implements `FromStr` this line won't be necessary anymore.

elsirion commented:
Also I forgot to mention: you can squash the first two and the last two commits. That way there is one changing the module API and one making use of it. Currently it's a bit too granular/the commits only make sense in each others context.

dpc commented on modules/fedimint-wallet-client/src/lib.rs:
```diff
@@ -437,6 +438,7 @@ impl ClientModuleGen for WalletClientGen {
 
     async fn init(
         &self,
+        _federation_id: FederationId,
```
Seems a bit one-off.

We keep tweaking the arguments to this function...

shaurya947 commented:
I think I've addressed all the feedback in the latest push. `rust-tests` passes  but `load-test-tool-test` fails with this error:

![image](https://github.com/fedimint/fedimint/assets/3454081/ed356149-11cc-4dd1-a27e-217137e37132)

To debug, I modified the below code to print the note:

![image](https://github.com/fedimint/fedimint/assets/3454081/5c04a485-f63a-4215-9d5a-3442eee5f67a)

I can see the notes being printed:

![image](https://github.com/fedimint/fedimint/assets/3454081/eb9b5a76-54ef-440d-83f1-72efbb542e11)

I tried the value in an online validator and indeed it's not padded (even though it's valid). I may be missing something small but how do I ensure that the encoding is padded as necessary?

shaurya947 commented:
I wonder if I'm doing this wrong by using `serde_as_encodable_hex` because that's hexifying the bytes after encoding them, and that's not the kind of string input we expect for `OOBNotes`. Going to try with custom ser/de impl to verify.

shaurya947 commented:
> I wonder if I'm doing this wrong by using `serde_as_encodable_hex` because that's hexifying the bytes after encoding them, and that's not the kind of string input we expect for `OOBNotes`. Going to try with custom ser/de impl to verify.

Ok, yes, that was it. I feel silly, hah! Should be all good now—I added a custom ser/de impl.

On second thought I suppose we _could_ use hex (and change the `FromStr` impl to account for it). Let me know if that's preferable and I can change it back.

elsirion commented:
Hex encoding would be significantly longer, that's why I chose base64 in the first place.

elsirion reviewed (approved):
Looks good, just needs a rebase :)

dpc commented:
Could have made `serde_as_encodable_base64` macro. :D

shaurya947 commented:
Thanks for the guidance folks, I enjoyed working on this enhancement :raised_hands:
